cedar-lean/Cedar/Partial/Response.lean

```diff
@@ -0,0 +1,162 @@
+/-
+ Copyright Cedar Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Data.Set
+import Cedar.Partial.Evaluator
+import Cedar.Partial.Expr
+import Cedar.Spec.Policy
+
+/-!
+This file defines Cedar partial responses.
+-/
+
+namespace Cedar.Partial
+
+open Cedar.Data
+open Cedar.Spec (Effect Error PolicyID)
+
+inductive Residual where
+  | residual (id : PolicyID) (effect : Effect) (condition : Partial.Expr)
+  | error (id : PolicyID) (error : Error) -- definitely results in this error, for any substitution of the unknowns
+
+deriving instance Repr, DecidableEq, Inhabited for Residual
+
+def Residual.id : Residual → PolicyID
+  | .residual id _ _ => id
+  | .error id _ => id
+
+def Residual.effect : Residual → Option Effect
+  | .residual _ effect _ => effect
+  | .error _ _ => none
+
+/--
+  if this policy must be satisfied (for any substitution of the unknowns), and
+  has the specified effect, return the PolicyID
+-/
+def Residual.mustBeSatisfied (eff : Effect) : Residual → Option PolicyID
+  | .residual id eff' (.lit (.bool true)) => if eff = eff' then some id else none
+  | _ => none
+
+/--
+  if this policy may be satisfied (for some substitution of the unknowns), and
+  has the specified effect, return the PolicyID
+-/
+def Residual.mayBeSatisfied (eff : Effect) : Residual → Option PolicyID
```

Comment:
This implementation is sound, but it could be more precise (unless I'm missing something).  In particular, we can return `.none` when the `condition` is the constant `false`.  Something along the lines of:

```
  | .residual id eff' (.lit (.bool false)) => none
  | .residual id eff' _ => if eff = eff' then some id else none
  | _ => none
```

I'm guessing we don't do this because the rest of the code never calls `.mayBeSatisfied` on a `false` policy due to some implicit invariants?

If so, it would be good to add the invariant to the comment above.

Reply:
However, that invariant only applies to `Response`, and you are right that this method could be used on a `Residual` outside of `Response`.  This is a good change

Reply:
There's no harm adding the line, even if we never call this on a residual with constant `false` in practice.